07.ls_object/ls.rb

```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/env ruby
+
+# frozen_string_literal: true
+
+require './ls_file'
+require 'optparse'
+
+class LS
+  MAX_ROW = 3
+
+  def main
+    opt = OptionParser.new
+
+    params = { dot_match: false, long_format: false, reverse: false }
+    opt.on('-a') { |v| params[:dot_match] = v }
+    opt.on('-l') { |v| params[:long_format] = v }
+    opt.on('-r') { |v| params[:reverse] = v }
+    opt.parse!(ARGV)
+    path = '.'
+
+    collected_filenames = collect_files(path, params[:dot_match], params[:reverse])
+    puts params[:long_format] ? list_long(collected_filenames) : list_short(collected_filenames)
+  end
+
+  private
+
+  def collect_files(path, dot_match, reverse)
+    all_filenames = Dir.foreach(path)
+    filenames = dot_match ? all_filenames : all_filenames.reject { |file| file.start_with?('.') }
+    sorted_filenames = sort_filenames(filenames)
+    reverse ? sorted_filenames.reverse : sorted_filenames
+  end
+
+  def sort_filenames(filenames)
+    filenames.sort do |a, b|
+      a = a.slice(1..-1) if a.start_with?('.') && !['.', '..'].include?(a)
+      b = b.slice(1..-1) if b.start_with?('.') && !['.', '..'].include?(b)
+      a.downcase <=> b.downcase
+    end
+  end
+
+  def list_long(filenames)
+    ls_file_objs = filenames.map { |filename| LSFile.new(filename) }
+    block_total = ls_file_objs.sum(&:blocks) / 2
+    total = "total #{block_total}"
+    body = format_body(ls_file_objs)
+    [total, *body].join("\n")
+  end
+
+  def format_body(ls_file_objs)
+    max_sizes = %i[nlink user group size].map do |key|
+      ls_file_objs.map { |ls_file| key_size(ls_file, key.to_s) }.max
+    end
+    ls_file_objs.map { |ls_file| format_row(ls_file, *max_sizes) }
+  end
+
+  def key_size(ls_file, key)
+    case key
+    when 'nlink'
+      ls_file.nlink.to_s.size
+    when 'user'
+      ls_file.user.size
+    when 'group'
+      ls_file.group.size
+    when 'size'
+      ls_file.file_size.to_s.size
+    when 'mtime'
+      ls_file.mtime.strftime('%b %e %H:%M').size
+    when 'blocks'
+      ls_file.blocks.size
+    end
+  end
+
+  def format_row(ls_file, max_nlink, max_user, max_group, max_size)
+    [
+      "#{ls_file.type}""#{ls_file.mode}",
```

Comment:
```suggestion
      "#{ls_file.type}#{ls_file.mode}",
```

Rubyの文法上、元の書き方でも動きますが、滅多に見かけない記法ですね

参考 https://qiita.com/jnchito/items/253b788301a6351a7b84

Reply:
修正しました